### DISCUSSION TITLE
Build system

### DISCUSSION THREAD

**Discussion by chiragsakhuja** [ChampSim/ChampSim#348] [Ideas]
I'm curious what constraints led to the design choices for the build system.  It seems contrived to use Python to generate C++ and then make some of that code transparent via defines passed into the compiler command.

I have a couple of ideas.

## Approach 1
Make all policies inherit from an abstract policy class, such as ReplacementPolicy.  Wherever needed, a ReplacementPolicy object could be instantiated. 

Pros:

- There wouldn't need to be any code generation at all.
- Policies could be selected dynamically at the command line.

Cons:

- The policy-writers would need to implement some sort of registration function so that the command line string of the policy name is mapped to the proper constructor (this could be hidden behind a convenient C-style macro).
- Compile times would increase if all policies are compiled. Alternatively, if dynamic policy selection isn't important, this could be remedied by only building/linking the single file that's specified in the JSON configuration as the current approach does.
- There would be some slight runtime overhead due to vtable accesses.

## Approach 2
All policies implement a standard interface, and any object that requires a policy could be templated.  For example, `CACHE<typename Repl, typename Pref>`, and functions could be called, for example, as follows `Repl::find_victim(...)` if the interface is fully static or, for example, `Repl`, if the interface uses OOP.  In theory, the interface could be enforced via Concepts in C++20, but this change probably doesn't warrant switching to an updated standard.

Pros:
- Greatly simplified build system, but there would still need to be some code generation (or at the very least, defines passed in through the command line) to instantiate the appropriate policy from the JSON config.
- Compared to Approach 1, there would be no runtime overhead.

Cons:
- The OOP version of this approach would require the policy writer to implement an entire class, which is a small increase in boilerplate code compared to the current approach.  The static version of this approach has a similar interface to the current approach, but it incentivizes the policy writer to use global or static variables, which can lead to messy code.

## Final Thoughts
Both approaches would require policy writers to update their code, and they just had to do that for the most recent ChampSim update.  Realistically, I don't expect any change in the near future, but I wanted to share the ideas in case they become relevant in the future.  I haven't fully thought through them, so I likely missed some pros/cons and implications.  Also, I don't have a very thorough understanding of the existing build system anyway (I am genuinely curious about the decisions that lead to it, though), so both approaches may be completely irrelevant/tone-deaf, in which case I apologize for imposing.

**Comment by ngober** [comment]
This is very insightful and absolutely correct in its reasoning. You've managed to discern a lot of the thoughts that have gone through my head over the past few years. In summary: the build system evolved out of a desire for minimal and incremental changes. There is a lot of legacy code available out on the internet, and we've wanted to maintain as much compatibility as we can, even as we have in fact broken compatibility in places. You can see much of the history in #69 and the links contained therein.

Concerning the particulars, the primary limitation is the legacy module architecture. Since the beginning of ChampSim, modules have been implementations of certain hook functions that are called by the core of the simulation. Early on, the build script would copy the specified modules, giving them a predefined name, and the whole thing would be compiled. This was inflexible and did not do well with incremental compilation (with make). In order to enable incremental compilation, we moved to a system where a makefile was generated that listed the actual files for each module to compile.

Then, when we moved to a more flexible cache layout following #124 and #129, another issue had to be solved: The module hooks had to coexist in the same `O3_CPU` and `CACHE` objects. Still to this point, hooks are member functions of their respective objects. The solution that I came to was name mangling with preprocessor defines. This required our build script to generate mangled hook signatures and functions that discriminate between the specified modules. This is the current state of the master branch, where the module being used is specified as a bitset passed to the constructor of the `O3_CPU` or `CACHE` objects.

As development continues, the architecture of the module system has continued to evolve. The module hooks are still name mangled, but the bitsets have disappeared. The module discriminators are currently implemented as type-erased member objects of the top-level objects. This is basically just a manually-constructed vtable that doesn't require inheritance. An upcoming change in #330 would start moving towards your Approach 2. In it, modules are defined as objects that implement certain functions. Those functions are constexpr-detected, so they don't have to implement all of the hooks, and acceptable defaults can be provided. This still requires some code generation, since we have to generate a makefile, but it's moving in a better direction. Also, just last week I merged #309 which would enable runtime parsing of the configuration file, so much more could be moved to runtime.

I appreciate you bringing this up and I would absolutely welcome your input on the future direction of the build system. You seem to have put a lot of thought into this.

**Comment by chiragsakhuja** [comment]
Thanks for the info.  I've looked through the pull requests you've linked and have developed a better understanding of the existing build system and the constraints that led to it.

## The issue
I've been playing around with a toy example to try out some ideas, and I think I've distilled the fundamental reason that the Python name-mangling is necessary in the existing build system.  To enable multiple hooks to co-exist in the same `CACHE` objects, you've essentially used Python to template the `CACHE` class.  In theory, this should be a fairly straightforward use case of templates or inheritance.  However, to use either of these approaches, the function/class definitions must be accessible to the compilation unit (e.g. `src/cache.cc`).  The module writers currently implement all their code in a single `.cc` file, so to make the definitions accessible to the compilation unit, either this entire file needs to be included in the compilation unit, or there needs to be some sort of glue code generation (the current approach).  The former is the more clean approach, but it gets in the way of incremental compilation because any change to the source of any module would trigger a rebuild of all included modules.

## A solution
The solution is pretty straightforward: require the module-writers to write separate header and source files.  I've attached a toy example that could be used as inspiration the next time you're willing to make a compatibility-breaking change.  The structure of this example is based on Approach 1 from my original post.  I tried out Approach 2 as well, but it required the `CACHE` class to be templated, which means every module would need to template the `CACHE` object every time it was referenced.  That seemed like a pretty unacceptable compromise just to avoid the small performance penalty of a vtable access, which is currently how things work anyway.

Note that in my toy example, I build all replacement policies rather than just the configured ones.  Though there is some initial overhead, I personally would prefer this behavior.  I don't feel strongly about it, so you could stick with the current approach of only building the relevant files.  In any case, I think my approach of using Python to tweak a Makefile.template may be a little easier for end users to understand (or at least be easier to display with syntax highlighting).  One concession I've made in terms of code generation is that I produce a `policies.inc` file that enumerates all the header files that should be included.

Disclaimer: I've only built the code and haven't checked whether my implemented policies/cache model behave as expected (though that part is irrelevant to the conversation anyway).

To build the example, run the following commands:
```
./configure
make
```

### Differences from existing system
I tried to minimize the differences that are visible to the module-writers, so it should be a pretty mindless task to do the conversion.  I believe the following is everything that would need to change in exiting modules.

1. The module implementations are in a virtual class now, so the module-writer has to write some extra, fairly simple, boiler-plate code.  There are a couple of advantages that come from writing modules in classes (discussed in the next section), so it may be a worthwhile change regardless.
2. The class and class members are defined in a separate header file now.
3. The function hooks are no longer part of the `CACHE` object, so the `CACHE` object must instead be passed into the hooks.  Existing modules would need to explicitly add `cache.*` to whichever members they were previously accessing.

### Tradeoffs

Pros:

1. Aside from the simple `policies.inc` file and the Makefile that get generated, there is no other file generation.  This feels like a cleaner and easier to understand approach.
2. There is no manual name mangling anymore.  All policies are controlled via a couple of defines that get inserted into the Makefile template when the configure script generates the Makefile.
3. Modules are isolated from their encapsulating classes.  This is a step in the right direction for a competition environment because the internals of the encapsulating classes can no longer be written to.  They can still be read from, but the only way around that is to only pass into the module the relevant `CACHE` members as a (potentially long) list of explicit parameters.
4. Modules no longer need to keep awkward state, such as the `lookahead` map in `prefetcher/ip_stride/ip_stride.cc`, to disambiguate the `CACHE` that it's associated with.
5. Because modules are wrapped in classes now, there is almost no risk of namespace collisions.
6. Because modules are wrapped in classes now, there is no longer a need to maintain global/static state in each module.  I think this is generally a more intuitive approach.
7. For the most part, this approach supports incremental compilation.

Cons:

1. The module-writers have to write some more code.  However, with my build system setup I've still limited the code modifications to just a single header and source file.  New modules get picked up and built automatically.
2. Whenever a module header file is changed, there is slightly more re-compilation than before.  All of the header files in `policies.inc` will need to be re-compiled.  However, they should only contain class definitions and not execution code, so it shouldn't make a big difference.

## Final thoughts
Hopefully my approach fits all the constraints of the build system!  Given how this flow works, SConstruct may be a better build system, but I've stuck with make.  One final consideration for an even more future version of ChampSim is to build all the modules into a single binary so that the policy can be selected at the command line (or potentially even during runtime).  I know that's something my lab would appreciate.  Switching over to my proposed build system already gets us through most of the issues hindering dynamic module selection.

[champsim-mvp.tar.gz](https://github.com/ChampSim/ChampSim/files/11639434/champsim-mvp.tar.gz)

**Comment by ngober** [comment]
Your suggestion is interesting, perhaps you should submit it as a pull request so that we could actually speak in terms of implementation. It does seem to be something close to what #330 is doing. However, in that patch, the modules are non-virtual subclasses and their member functions are examined at compile time. It probably comes out to approximately the same thing.

I'm very open to an idea of runtime module selection. It would be great to move configuration processing into the runtime, but the user still needs to provide compile paths.
